tests/resolver/test_filtering.py

```diff
@@ -0,0 +1,405 @@
+from __future__ import annotations
+
+import copy
+import random
+from collections import deque
+
+import pytest
+
+from variantlib.errors import ValidationError
+from variantlib.models.variant import VariantDescription
+from variantlib.models.variant import VariantFeature
+from variantlib.models.variant import VariantProperty
+from variantlib.resolver.filtering import filter_variants_by_features
+from variantlib.resolver.filtering import filter_variants_by_namespaces
+from variantlib.resolver.filtering import filter_variants_by_property
+from variantlib.resolver.filtering import remove_duplicates
+
+
+@pytest.fixture(scope="session")
+def vprops() -> list[VariantProperty]:
+    return [
+        VariantProperty(namespace="OmniCorp", feature="custom_feat", value="value1"),
+        VariantProperty(
+            namespace="TyrellCorporation", feature="client_id", value="value2"
+        ),
+    ]
+
+
+@pytest.fixture(scope="session")
+def vdescs(vprops: list[VariantProperty]) -> list[VariantDescription]:
+    """Fixture to create a list of VariantDescription objects."""
+    assert len(vprops) == 2
+    prop1, prop2 = vprops
+
+    return [
+        VariantDescription([prop1]),
+        VariantDescription([prop2]),
+        VariantDescription([prop1, prop2]),
+    ]
+
+
+# =========================== `remove_duplicates` =========================== #
+
+
+def test_remove_duplicates(vdescs: list[VariantDescription]):
+    # using `copy.deepcopy` to ensure that all objects are actually unique
+    input_vdescs = [copy.deepcopy(random.choice(vdescs)) for _ in range(100)]
+    filtered_vdescs = list(remove_duplicates(input_vdescs))
+
+    assert len(filtered_vdescs) == 3
+
+    for vdesc in vdescs:
+        assert vdesc in filtered_vdescs
```

Comment:
Wouldn't

```suggestion
    assert sorted(filtered_vdescs) == sorted(vdescs)
```

work? Comparing lists directly should give better diffs.

Reply:
Does not work because `VariantDescription` does not support ordering. And it's a good thing. The whole logic of this PR is to provide a sorting interface conditional of a variant provider sorted list.
Adding a naive ordering could be confusing (and lead to silent bugs).